internal/db/p2p_replicator.go

```diff
@@ -0,0 +1,341 @@
+			if err != nil {
+				log.ErrorContextE(
+					ctx,
+					"Failed to get all docIDs",
+					NewErrReplicatorDocID(err, errors.NewKV("Collection", col.Name().Value())),
+				)
+				continue
+			}
+			for docIDResult := range keysCh {
+				if docIDResult.Err != nil {
+					log.ErrorContextE(ctx, "Key channel error", docIDResult.Err)
+					continue
+				}
+				docID := core.DataStoreKeyFromDocID(docIDResult.ID)
+				headset := clock.NewHeadSet(
+					txn.Headstore(),
+					docID.WithFieldId(core.COMPOSITE_NAMESPACE).ToHeadStoreKey(),
+				)
+				cids, _, err := headset.List(ctx)
+				if err != nil {
+					log.ErrorContextE(
+						ctx,
+						"Failed to get heads",
+						err,
+						corelog.String("DocID", docIDResult.ID.String()),
+						corelog.Any("Collection", col.Name()))
+					continue
+				}
+				// loop over heads, get block, make the required logs, and send
+				for _, c := range cids {
+					blk, err := txn.Blockstore().Get(ctx, c)
+					if err != nil {
+						log.ErrorContextE(ctx, "Failed to get block", err,
+							corelog.Any("CID", c),
+							corelog.Any("Collection", col.Name()))
+						continue
+					}
+
+					updateChan <- event.Update{
+						DocID:      docIDResult.ID.String(),
+						Cid:        c,
+						SchemaRoot: col.SchemaRoot(),
+						Block:      blk.RawData(),
+					}
+				}
+			}
+		}
+	}()
+
+	return updateChan
+}
+
+func (db *db) DeleteReplicator(ctx context.Context, rep client.Replicator) error {
+	txn, err := db.NewTxn(ctx, false)
+	if err != nil {
+		return err
+	}
+	defer txn.Discard(ctx)
+
+	if err := rep.Info.ID.Validate(); err != nil {
+		return err
+	}
+
+	// set transaction for all operations
+	ctx = SetContextTxn(ctx, txn)
+
+	oldSchemas := make(map[string]struct{})
+	repKey := core.NewReplicatorKey(rep.Info.ID.String())
+	hasOldRep, err := txn.Systemstore().Has(ctx, repKey.ToDS())
+	if err != nil {
+		return err
+	}
+	if hasOldRep {
+		repBytes, err := txn.Systemstore().Get(ctx, repKey.ToDS())
+		if err != nil {
+			return err
+		}
+		oldRep := client.Replicator{}
+		err = json.Unmarshal(repBytes, &oldRep)
+		if err != nil {
+			return err
+		}
+		for _, schema := range oldRep.Schemas {
+			oldSchemas[schema] = struct{}{}
+		}
+	}
+
+	var collections []client.Collection
+	if len(rep.Schemas) > 0 {
+		// if specific collections are chosen get them by name
+		for _, name := range rep.Schemas {
+			col, err := db.GetCollectionByName(ctx, name)
+			if err != nil {
+				return NewErrReplicatorCollections(err)
+			}
+			collections = append(collections, col)
+		}
+		// make sure the replicator exists in the datastore
+		key := core.NewReplicatorKey(rep.Info.ID.String())
+		_, err = txn.Systemstore().Get(ctx, key.ToDS())
+		if err != nil {
+			return err
+		}
+	} else {
+		oldSchemas = make(map[string]struct{})
+	}
+
+	rep.Schemas = nil
+	for _, col := range collections {
+		delete(oldSchemas, col.SchemaRoot())
+	}
+	for schema := range oldSchemas {
+		rep.Schemas = append(rep.Schemas, schema)
+	}
```

Comment:
suggestion: If new code, one liner comment would be useful for what/why of this code.

Reply:
is covered by integration test but now also covered by unit test